Improve resin-filesystem-expand script

The `return 1` statement in the following script gives an error in case there is a timeout in the appearance of a file-system label for any reason (sd card/storage corruption usually)

https://github.com/resin-os/meta-resin/blob/ff0e3716e58c48ecd31a5e46d13a68a880507518/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand/resin-filesystem-expand#L9-L16

The error

```
May 23 12:55:50 xxxxxxx sh[495]: [ERROR] Timeout while waiting for filesystem label to appear.
May 23 12:55:50 xxxxxxx sh[495]: /usr/bin/resin-filesystem-expand: line 15: return: can only `return' from a function or sourced script
```

In case the file-system labels don't appear in 30 seconds, assume something is buggered and attempt to recover the file-system by running fsck on everything!

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_bzqj)

--- a/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand.bb
+++ b/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand.bb
@@ -15,6 +15,7 @@ SYSTEMD_SERVICE_${PN} = "resin-filesystem-expand.service"
 RDEPENDS_${PN} = " \
     coreutils \
     e2fsprogs-resize2fs \
+    util-linux \
     "
 
 do_install() {

--- a/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand/resin-filesystem-expand
+++ b/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand/resin-filesystem-expand
@@ -11,7 +11,16 @@ while [ ! -L "$DATA_PARTITION" ]; do
 		sleep 1 && end=$((end + 1))
 	else
 		echo "[ERROR] Timeout while waiting for filesystem label to appear."
-		return 1
+		if [ ! -f /mnt/boot/resin_attempting_fsck_recovery ]; then
+			echo "[ERROR] Something bad happened. Assuming storage corruption and attempting fsck and reboot."
+			touch /mnt/boot/resin_attempting_fsck_recovery
+			fdisk -l | grep '^/dev' | grep -i 'FAT' |  cut -d' ' -f1 | while read line ; do fsck.fat -av $line; done
+			fdisk -l | grep '^/dev' | grep -i 'Linux' |  cut -d' ' -f1 | while read line ; do fsck.ext4 -pfv $line; done
+			/bin/systemctl --force reboot
+		else
+			echo "[ERROR] Something bad happened. Failed at recovering via an fsck and reboot cycle"
+			exit 0
+		fi
 	fi
 done
 
@@ -20,5 +29,10 @@ e2fsck -y $DATA_PARTITION
 echo "[INFO] Expand ext4 filesystem on $DATA_PARTITION... "
 resize2fs -f $DATA_PARTITION
 sync
+
+if [ -f /mnt/boot/resin_attempting_fsck_recovery ]; then
+	echo "[WARN] Looks like there was a successful automated fsck recovery. Please check the integrity of your storage device"
+	rm /mnt/boot/resin_attempting_fsck_recovery
+fi
 
 exit 0

--- a/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand/resin-filesystem-expand.service
+++ b/meta-resin-common/recipes-support/resin-expand/resin-filesystem-expand/resin-filesystem-expand.service
@@ -2,6 +2,7 @@
 Description=Expand resin-data partition
 DefaultDependencies=no
 Before=resin-data.service umount.target
+After=resin-boot.service
 Conflicts=umount.target
 ConditionVirtualization=!docker
 

